cv32e40s/tests/programs/custom/minhv_pma_block/minhv_pma_block.c

```diff
@@ -0,0 +1,174 @@
+//
+// Copyright 2023 Silicon Labs, Inc.
+//
+// Licensed under the Solderpad Hardware Licence, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://solderpad.org/licenses/
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+///////////////////////////////////////////////////////////////////////////////
+//
+// Author: Kristine Døsvik
+//
+// Test minhv=1 and pma block of mepc address.
+//
+/////////////////////////////////////////////////////////////////////////////////
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdint.h>
+
+
+extern volatile uint32_t recovery_pt_mret;
+volatile uint32_t g_entered_trap_handler;
+volatile uint32_t g_trapped_mcause;
+
+int execute_from_io_region();
+__attribute__((interrupt ("machine"))) void u_sw_irq_handler(void);
+void trap_handler(void);
+
+
+int main(int argc, char **argv){
+
+  int is_failure = 0;
+
+  is_failure += execute_from_io_region();
+
+  if(is_failure){
+    return EXIT_FAILURE;
+  } else {
+    return EXIT_SUCCESS;
+  }
+}
+
+
+int execute_from_io_region() {
+
+  volatile uint32_t io_addr = 0xFFFFEFFF << 2;
+  volatile uint32_t mcause_excode = 0x0;
+  volatile uint32_t mcause_excode_expected = 0x1; //Execution attempt from I/O region.
+  volatile uint32_t mret_dont_trap = 0;
+  volatile uint32_t test_fail = 0;
+  g_entered_trap_handler = 0;
+
+  // Load an IO memory address into mepc
+  __asm__ volatile ( R"(
+    csrrw zero, mepc, %[io_addr]
+  )" :: [io_addr] "r"(io_addr));
+
+  // Set minhv and mpp = M
+  __asm__ volatile ( R"(
+    lui t0, 0x70000
+    csrrs zero, mcause, t0
+  )"::: "t0");
+
+  __asm__ volatile ( R"(
+    mret
+  )":::);
+
+  // Mret traps.
+
+  // This should never execute (deliberate dead code), as the mret make the pc take a jump
+  mret_dont_trap += 1;
```

Comment:
Have we confirmed in the object dump that this doesn't get optimized away?

(I don't suspect that it does, but I can't reason that it is impossible.)

Reply:
Good point! I have looked through the object dump file. And the code seems to be constructed as expected.